src/engine/Service.cpp

```diff
@@ -504,3 +495,117 @@ std::optional<std::string> Service::idToValueForValuesClause(
       }
   }
 }
+
+// ____________________________________________________________________________
+void Service::precomputeSiblingResult(std::shared_ptr<Operation> left,
+                                      std::shared_ptr<Operation> right,
+                                      bool rightOnly, bool requestLaziness) {
+  AD_CORRECTNESS_CHECK(left && right);
+
+  auto skipSortOperation = [](std::shared_ptr<Operation>& op) {
+    auto children = op->getChildren();
+    if (children.size() == 1) {
+      op = children[0]->getRootOperation();
+    }
+  };
+  skipSortOperation(left);
+  skipSortOperation(right);
+
+  auto a = std::dynamic_pointer_cast<Service>(left);
+  auto b = std::dynamic_pointer_cast<Service>(right);
+
+  if ((rightOnly && !static_cast<bool>(b)) ||
+      (!rightOnly && static_cast<bool>(a) == static_cast<bool>(b))) {
+    return;
+  }
+
+  const auto& [service, sibling] = [&]() {
+    if (a) {
+      return std::tie(a, right);
+    } else {
+      AD_CORRECTNESS_CHECK(b);
+      return std::tie(b, left);
+    }
+  }();
+
+  auto addRuntimeInfo = [&](bool siblingUsed) {
+    std::string_view v = siblingUsed ? "yes"sv : "no"sv;
+    service->runtimeInfo().addDetail("optimized-with-sibling-result", v);
+    sibling->runtimeInfo().addDetail("used-to-optimize-service-sibling", v);
+  };
+
+  auto siblingResult = sibling->getResult(
+      false, requestLaziness ? ComputationMode::LAZY_IF_SUPPORTED
+                             : ComputationMode::FULLY_MATERIALIZED);
+
+  if (siblingResult->isFullyMaterialized()) {
+    sibling->precomputedResultBecauseSiblingOfService_ = siblingResult;
```

Comment:
You could do this step as the second one, then you can `std::move(...)`  the `siblingResult`.